util/bloomchain/src/chain.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::{HashMap, HashSet};
+use std::ops::Range;
+use number::Number;
+use position::{Position, Manager as PositionManager};
+use bloom::Bloom;
+use filter::Filter;
+use config::Config;
+use database::BloomDatabase;
+
+/// Prepares all bloom database operations.
+pub struct BloomChain<'a> {
+	positioner: PositionManager,
+	db: &'a BloomDatabase,
+}
+
+impl<'a> BloomChain<'a> {
+	/// Creates new bloom chain.
+	pub fn new(config: Config, db: &'a BloomDatabase) -> Self {
+		let positioner = PositionManager::new(config.elements_per_index, config.levels);
+
+		BloomChain {
+			positioner: positioner,
+			db: db,
+		}
+	}
+
+	/// Internal function which does bloom search recursively.
+	fn blocks(&self, range: &Range<Number>, bloom: &Bloom, level: usize, offset: usize) -> Option<Vec<usize>> {
+		let index = self.positioner.position(offset, level);
+
+		match self.db.bloom_at(&index) {
+			None => return None,
+			Some(level_bloom) => match level {
+				// if we are on the lowest level
+				0 if level_bloom.contains_bloom(bloom) => return Some(vec![offset]),
+				// return None if current level doesnt contain given bloom
+				_ if !level_bloom.contains_bloom(bloom) => return None,
```

Comment:
Maybe swap with first arm (`0 if ...`) and leave single `contains_bloom` call? Like:
`
match level {
 _ if !level_bloom.contains(bloom) => ...
 0 => ...
 _ => ...
}
`